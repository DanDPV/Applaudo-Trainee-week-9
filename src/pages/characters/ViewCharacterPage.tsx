import React from 'react';

const ViewCharacterPage = () => <h1>View Character Page</h1>;

export default ViewCharacterPage;
