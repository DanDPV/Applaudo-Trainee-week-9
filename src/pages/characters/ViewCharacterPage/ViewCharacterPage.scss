$bg-title-image: 'https://wallpaperaccess.com/full/229965.jpg';
$red: #e74c3c;

:root {
  --font-sans: 'Rubik', sans-serif;
}

.image-header-div {
  color: white;
  height: 300px;
  display: flex;
  justify-content: center;
  align-items: flex-end;
  padding-bottom: 20px;

  background: -moz-linear-gradient(
      top,
      rgba(0, 0, 0, 0) 0%,
      rgba(0, 0, 0, 0) 59%,
      rgba(0, 0, 0, 0.65) 100%
    ),
    url($bg-title-image) no-repeat;
  background: -webkit-gradient(
      linear,
      left top,
      left bottom,
      color-stop(0%, rgba(0, 0, 0, 0)),
      color-stop(59%, rgba(0, 0, 0, 0)),
      color-stop(100%, rgba(0, 0, 0, 0.65))
    ),
    url($bg-title-image) no-repeat;
  background: -webkit-linear-gradient(
      top,
      rgba(0, 0, 0, 0) 0%,
      rgba(0, 0, 0, 0) 59%,
      rgba(0, 0, 0, 0.65) 100%
    ),
    url($bg-title-image) no-repeat;
  background: -o-linear-gradient(
      top,
      rgba(0, 0, 0, 0) 0%,
      rgba(0, 0, 0, 0) 59%,
      rgba(0, 0, 0, 0.65) 100%
    ),
    url($bg-title-image) no-repeat;
  background: -ms-linear-gradient(
      top,
      rgba(0, 0, 0, 0) 0%,
      rgba(0, 0, 0, 0) 59%,
      rgba(0, 0, 0, 0.65) 100%
    ),
    url($bg-title-image) no-repeat;
  background: linear-gradient(
      to bottom,
      rgba(0, 0, 0, 0) 0%,
      rgba(0, 0, 0, 0) 59%,
      rgba(0, 0, 0, 0.65) 100%
    ),
    url($bg-title-image) no-repeat;

  background-position: top;
}

.back-button {
  position: absolute;
  top: 15px;
  left: 15px;
  box-sizing: border-box;
  appearance: none;
  background-color: black;
  border: 2px solid $red;
  border-radius: 0.6em;
  color: $red;
  cursor: pointer;
  display: flex;
  justify-content: flex-end;
  font-size: 0.8rem;
  font-weight: 400;
  line-height: 1;
  padding: 1.3em 2em;
  text-decoration: none;
  text-transform: uppercase;
  font-family: 'Montserrat', sans-serif;
  font-weight: 700;

  &:hover,
  &:focus {
    color: #fff;
    outline: 0;
  }
}

.back-button-decoration {
  border-color: $red;
  border-radius: 0;
  color: $red;
  overflow: hidden;
  z-index: 1;
  transition: color 150ms ease-in-out;

  &:after {
    content: '';
    position: absolute;
    display: block;
    top: 0;
    left: 50%;
    transform: translateX(-50%);
    width: 0;
    height: 100%;
    background: $red;
    z-index: -1;
    transition: width 150ms ease-in-out;
  }

  &:hover {
    color: #fff;
    &:after {
      width: 110%;
    }
  }
}

.image-header-title {
  text-align: center;
}

.char-description {
  display: flex;
  flex-wrap: wrap;
  margin: 15px 15%;
}

.char-image {
  margin: 10px 10px 0px;
  height: 350px;
  width: 200px;
  min-width: 250px;
  background-size: contain;
  background-position: center;
  background-repeat: no-repeat;
}

.char-info {
  width: 65%;
  text-align: justify;
  margin-left: 5px;
  margin-top: 10px;
  display: flex;
  flex-direction: column;
  flex-wrap: wrap;
  justify-content: center;
}

.char-description-p {
  margin-bottom: 20px;
}

.char-extra-title {
  font-weight: bold;
  margin-bottom: 5px;
}

@media only screen and (max-width: 600px) {
  .back-button {
    padding: 0.8em 1em;
    font-size: 0.7rem;
  }

  .image-header-title {
    font-size: 23px;
    text-shadow: -1px 1px 0 #000, 1px 1px 0 #000, 1px -1px 0 #000,
      -1px -1px 0 #000;
    padding: 0 5px;
  }

  .image-header-div {
    height: 150px;
  }
}
