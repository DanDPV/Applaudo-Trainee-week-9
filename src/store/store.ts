import searchReducer from 'reducers/search/searchReducer';
import { createStore, combineReducers } from 'redux';

const rootReducer = combineReducers({
  search: searchReducer,
});

const store = createStore(rootReducer);

export default store;
