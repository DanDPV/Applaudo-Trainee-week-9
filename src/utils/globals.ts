/* eslint-disable import/prefer-default-export */
const imagePlaceholder = 'https://images7.alphacoders.com/514/thumb-1920-514639.jpg';

export { imagePlaceholder };
