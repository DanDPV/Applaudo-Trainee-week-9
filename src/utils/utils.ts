/* eslint no-restricted-syntax: 0 */
const getQueryVariable = (variable: string) => {
  const query = window.location.search.substring(1);
  const vars = query.split('&');

  for (const v of vars) {
    const pair = v.split('=');
    if (pair[0] === variable) {
      return pair[1];
    }
  }

  return '';
};

const shortenText = (text: string, length: number) => {
  if (text.length <= length) {
    return text;
  }
  let sub = text.substr(0, length);
  sub += '...';
  return sub;
};

export { getQueryVariable, shortenText };
