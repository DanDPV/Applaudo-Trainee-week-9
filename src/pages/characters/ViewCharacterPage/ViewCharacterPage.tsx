import React, { useEffect, useState } from 'react';
import queryString from 'query-string';
import { useDispatch, useSelector } from 'react-redux';
import { useHistory, useParams } from 'react-router-dom';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faChevronLeft } from '@fortawesome/free-solid-svg-icons';
import { get } from 'API/FetchInfo';
import { IRootState } from 'store/store';
import { setViewItemAsyncContent } from 'actions/viewItem';
import ICharacter from 'interfaces/ICharacter';
import IGenericApiResponse from 'interfaces/IGenericApiResponse';
import Loading from 'components/Loading/Loading';
import CustomOrderedList from 'components/CustomOrderedList/CustomOrderedList';
import { imagePlaceholder } from 'utils/globals';
import RouteNames from 'routers/RouteNames';
import 'pages/characters/ViewCharacterPage/ViewCharacterPage.scss';

const ViewCharacterPage = () => {
  interface pathParams {
    idCharacter: string;
  }

  const history = useHistory();
  const dispatch = useDispatch();
  const { idCharacter } = useParams<pathParams>();
  const {
    loading,
    error,
    data: genericResponse,
  } = useSelector((state: IRootState) => state.viewItem);

  const { data } = genericResponse ?? {};
  const { results } = data ?? {};
  const [character, setCharacter] = useState<ICharacter | null>(null);

  const handleBack = () => {
    if (history.length <= 2) history.push(RouteNames.Home);
    else history.goBack();
  };

  const handleViewComic = (id: string) => history.push(`/comics/${id}`);

  const handleViewStory = (id: string) => history.push(`/stories/${id}`);

  useEffect(() => {
    if (results) {
      if (!('characters' in results[0])) {
        setCharacter((results[0] as unknown) as ICharacter);
      }
    }
  }, [results]);

  useEffect(() => {
    dispatch(setViewItemAsyncContent(true, '', null));
    const url = `${process.env.REACT_APP_API_URL}v1/public/characters/${idCharacter}?${queryString.stringify({
      apikey: process.env.REACT_APP_PUBLIC_KEY,
    })}`;
    get<IGenericApiResponse<ICharacter>>(url)
      .then(res => {
        dispatch(setViewItemAsyncContent(false, '', res));
      })
      .catch(() => {
        dispatch(setViewItemAsyncContent(false, 'Could not load character', null));
      });
  }, []);

  return (
    <div className="main-content mb-5">
      {loading && <Loading />}
      {error && <h2 className="error-message">Could not load character ðŸ˜“</h2>}
      {!loading && !error && character && (
        <>
          <div className="image-header-div">
            <button
              type="button"
              className="back-button back-button-decoration"
              onClick={handleBack}
            >
              <FontAwesomeIcon icon={faChevronLeft} />
              {'\u00A0'}
              Back
            </button>
            <h1 className="image-header-title">{character.name}</h1>
          </div>
          <div className="char-description">
            <div
              className="char-image"
              style={{
                backgroundImage: `url('${
                  character.thumbnail
                    ? `${character.thumbnail.path}/portrait_uncanny.${character.thumbnail.extension}`
                    : imagePlaceholder
                }')`,
              }}
            />
            <div className="char-info">
              <p className="char-description-p">{character.description}</p>
              {character.urls && character.urls.length > 0 && (
                <>
                  <p className="char-extra-title">
                    {`Learn more about ${character.name}`}
                  </p>
                  <div className="char-extra-div">
                    {character.urls.map(res => (
                      <a
                        className="char-extra-button"
                        target="_blank"
                        rel="noopener noreferrer"
                        key={res.type}
                        href={res.url}
                      >
                        {res.type}
                      </a>
                    ))}
                  </div>
                </>
              )}
            </div>
          </div>
          <div className="related-items-div">
            <h3 className="related-items-title">Comics</h3>
            {character.comics.returned > 0
              ? <CustomOrderedList items={character.comics.items} onClick={handleViewComic} />
              : <p className="not-found">No comics found ðŸ¤”</p>}
          </div>
          <div className="related-items-div">
            <h3 className="related-items-title">Stories</h3>
            {character.stories.returned > 0
              ? <CustomOrderedList items={character.stories.items} onClick={handleViewStory} />
              : <p className="not-found">No stories found ðŸ¤”</p>}
          </div>
        </>
      )}
    </div>
  );
};

export default ViewCharacterPage;
