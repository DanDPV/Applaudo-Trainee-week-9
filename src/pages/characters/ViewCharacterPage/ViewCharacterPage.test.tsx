/* eslint-disable import/extensions */
/* eslint-disable no-undef */
import React from 'react';
import { Provider } from 'react-redux';
import { screen, waitFor } from '@testing-library/react';
import store from 'store/store';
import { renderWithRouter } from 'tests/utils';
import ViewCharacterPage from './ViewCharacterPage';

describe('Test in ViewCharacterPage component', () => {
  test('should render component correctly', async () => {
    const { container } = renderWithRouter(
      <Provider store={store}>
        <ViewCharacterPage />
      </Provider>, { route: '/characters/1' },
    );

    await waitFor(() => {
      expect(container.querySelector('.image-header-title')).not.toBeNull();
      expect(container.querySelector('.image-header-title')?.innerHTML).toBe('spiderman');
      expect(container).toMatchSnapshot();
    });
  });

  test('should show comic and story list correctly', async () => {
    const { container } = renderWithRouter(
      <Provider store={store}>
        <ViewCharacterPage />
      </Provider>,
      { route: '/characters/1' },
    );

    await waitFor(() => {
      expect(container.querySelector('.image-header-title')).not.toBeNull();
    });

    expect(screen.getByText(/Spiderman Comic/i)).toBeInTheDocument();
    expect(screen.getByText(/Spiderman Story/i)).toBeInTheDocument();
  });
});
