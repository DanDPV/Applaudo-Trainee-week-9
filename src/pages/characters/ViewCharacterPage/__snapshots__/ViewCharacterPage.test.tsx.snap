// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test in ViewCharacterPage component should render component correctly 1`] = `
<div>
  <div
    class="main-content mb-5"
  >
    
    <div
      class="image-header-div"
    >
      <button
        class="back-button back-button-decoration"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-chevron-left fa-w-10 "
          data-icon="chevron-left"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 320 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M34.52 239.03L228.87 44.69c9.37-9.37 24.57-9.37 33.94 0l22.67 22.67c9.36 9.36 9.37 24.52.04 33.9L131.49 256l154.02 154.75c9.34 9.38 9.32 24.54-.04 33.9l-22.67 22.67c-9.37 9.37-24.57 9.37-33.94 0L34.52 272.97c-9.37-9.37-9.37-24.57 0-33.94z"
            fill="currentColor"
          />
        </svg>
         
        Back
      </button>
      <h1
        class="image-header-title"
      >
        spiderman
      </h1>
    </div>
    <div
      class="char-description fade-in"
    >
      <div
        class="char-image"
        style="background-image: url(path/portrait_uncanny.jpg);"
      />
      <div
        class="char-info"
      >
        <p
          class="char-description-p"
        >
          description
        </p>
        <p
          class="char-extra-title"
        >
          Learn more about spiderman
        </p>
        <div
          class="char-extra-div"
        >
          <a
            class="char-extra-button"
            href="http://marvel.com/characters/2854/adam_warlock?utm_campaign=apiRef&utm_source=10dea460449717d1dce670f173fe30b9"
            rel="noopener noreferrer"
            target="_blank"
          >
            detail
          </a>
          <a
            class="char-extra-button"
            href="http://marvel.com/universe/Warlock,_Adam?utm_campaign=apiRef&utm_source=10dea460449717d1dce670f173fe30b9"
            rel="noopener noreferrer"
            target="_blank"
          >
            wiki
          </a>
        </div>
      </div>
    </div>
    <div
      class="actions-title-div"
    >
      <p
        class="actions-title"
      >
        Actions:
      </p>
    </div>
    <div
      class="actions-menu"
    >
      <button
        class="action-btn add-bookmark"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-bookmark fa-w-12 "
          data-icon="bookmark"
          data-prefix="far"
          focusable="false"
          role="img"
          viewBox="0 0 384 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M336 0H48C21.49 0 0 21.49 0 48v464l192-112 192 112V48c0-26.51-21.49-48-48-48zm0 428.43l-144-84-144 84V54a6 6 0 0 1 6-6h276c3.314 0 6 2.683 6 5.996V428.43z"
            fill="currentColor"
          />
        </svg>
         
        Add bookmark
      </button>
      <button
        class="action-btn hide-item"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-eye-slash fa-w-20 "
          data-icon="eye-slash"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 640 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M320 400c-75.85 0-137.25-58.71-142.9-133.11L72.2 185.82c-13.79 17.3-26.48 35.59-36.72 55.59a32.35 32.35 0 0 0 0 29.19C89.71 376.41 197.07 448 320 448c26.91 0 52.87-4 77.89-10.46L346 397.39a144.13 144.13 0 0 1-26 2.61zm313.82 58.1l-110.55-85.44a331.25 331.25 0 0 0 81.25-102.07 32.35 32.35 0 0 0 0-29.19C550.29 135.59 442.93 64 320 64a308.15 308.15 0 0 0-147.32 37.7L45.46 3.37A16 16 0 0 0 23 6.18L3.37 31.45A16 16 0 0 0 6.18 53.9l588.36 454.73a16 16 0 0 0 22.46-2.81l19.64-25.27a16 16 0 0 0-2.82-22.45zm-183.72-142l-39.3-30.38A94.75 94.75 0 0 0 416 256a94.76 94.76 0 0 0-121.31-92.21A47.65 47.65 0 0 1 304 192a46.64 46.64 0 0 1-1.54 10l-73.61-56.89A142.31 142.31 0 0 1 320 112a143.92 143.92 0 0 1 144 144c0 21.63-5.29 41.79-13.9 60.11z"
            fill="currentColor"
          />
        </svg>
         
        Hide item
      </button>
    </div>
    <div
      class="related-items-div fade-in"
    >
      <h3
        class="related-items-title"
      >
        Comics
      </h3>
      <ol
        class="custom-ol"
      >
        <li
          class="custom-li"
        >
          Spiderman Comic
        </li>
      </ol>
    </div>
    <div
      class="related-items-div fade-in"
    >
      <h3
        class="related-items-title"
      >
        Stories
      </h3>
      <ol
        class="custom-ol"
      >
        <li
          class="custom-li"
        >
          Spiderman Story
        </li>
      </ol>
    </div>
  </div>
</div>
`;
